From fa1a52bfdd22e414fc84fba9be170f9d1433996a Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 12 Dec 2024 18:18:42 +0800
Subject: [PATCH 081/107] mtk: mac80211: fix cross link csa issue

Fix cross link csa issue
1. Parsing elements in per-sta profile without skipping capability info
   This leads to CSA IE parsing failure for per-STA profile.
2. Handling radar detection during CSA & CAC period

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/mlme.c | 20 +++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index b90cb9a1..f19597c4 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -2732,11 +2732,6 @@ ieee80211_sta_process_chanswitch(struct ieee80211_link_data *link,
 			}
 			break;
 		case IEEE80211_CSA_SOURCE_OTHER_LINK:
-			/* active link: we want to see the beacon to continue */
-			if (ieee80211_vif_link_active(&sdata->vif,
-						      link->link_id))
-				return;
-
 			/* switch work ran, so just complete the process */
 			if (link->u.mgd.csa.waiting_bcn) {
 				ieee80211_chswitch_post_beacon(link);
@@ -2749,6 +2744,11 @@ ieee80211_sta_process_chanswitch(struct ieee80211_link_data *link,
 				break;
 			}
 
+			/* active link: we want to see the beacon to continue */
+			if (ieee80211_vif_link_active(&sdata->vif,
+						      link->link_id))
+				return;
+
 			/* link still has CSA but we already know, do nothing */
 			if (!res)
 				return;
@@ -7204,7 +7204,7 @@ ieee80211_mgd_check_cross_link_csa(struct ieee80211_sub_if_data *sdata,
 		struct ieee80211_mle_per_sta_profile *prof;
 		struct ieee802_11_elems *prof_elems;
 		struct ieee80211_link_data *link;
-		ssize_t len;
+		ssize_t len, capab_info_len = sizeof(u16);
 
 		if (link_id == reporting_link_id)
 			continue;
@@ -7228,10 +7228,12 @@ ieee80211_mgd_check_cross_link_csa(struct ieee80211_sub_if_data *sdata,
 			continue;
 
 		prof = (void *)sta_profiles[link_id];
+		/* skip capab info (u16) */
 		prof_elems = ieee802_11_parse_elems(prof->variable +
-						    (prof->sta_info_len - 1),
-						    len -
-						    (prof->sta_info_len - 1),
+						    (prof->sta_info_len - 1) +
+						    capab_info_len,
+						    len - (prof->sta_info_len - 1) -
+						    capab_info_len,
 						    false, NULL);
 
 		/* memory allocation failed - let's hope that's transient */
-- 
2.45.2

